src/storage/sstable.rs

```diff
@@ -179,4 +171,31 @@ impl SSTable {
 
         Ok(vec![t])
     }
+}
+
+#[derive(Debug)]
+struct SparseIndex {
+    // this version of the index is backed by an ordered map.
+    map: BTreeMap<Key, FileOffset>,
+}
+
+impl SparseIndex {
+    fn new() -> Self {
+        Self {
+            map: Default::default(),
+        }
+    }
+
+    fn insert(&mut self, key: Key, offset: FileOffset) {
```

Comment:
What was your motivation for separating out this struct? For example, `new()` and `insert()` can both be done via `#[derive(Default, Deref)]`. I'm aware we're going to replace the BTreeMap with a Vec. Were there other changes?

Reply:
For readability, clarity, and encapsulation. 

---

* (clarity) The intention is to provide clarity about the operations performed using the sparse index. It sacrifices raw line-count for the benefit that comes with defining an ADT:

> An ADT is a mathematical model of a data structure that specifies the type of data stored, the operations supported on them, and the types of parameters of the operations.  An ADT specifies what each operation does, but not how it does it.  Typically, an ADT can be implemented using one of many different data structures.  A useful first step in deciding what data structure to use in a program is to specify an ADT for the program.

* (encapsulation) When we swap out the `BTreeMap` for something else, the call-sites don't need to change. The change becomes isolated.

* (readability) The function signatures become an API contract. For instance, we know that we want the operations to have certain worst-case complexities. It becomes easier to eyeball the data structure and quickly inspect.  `insert` to be `O(1)` and `find_offset` to be O(1) potentially or at the very least: O(lgn).

* An aside: we may need to change from `insert(k, o)` to `insert_bulk(Vec<(k,o)>)` in order to have O(1) insertions if we intend to insert in order.

Reply:
Ok